indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamEndSequenceNumbers.java

```diff
@@ -0,0 +1,214 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.seekablestream;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import org.apache.druid.java.util.common.IAE;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+
+/**
+ * Represents the end sequenceNumber per partition of a sequence. Note that end sequenceNumbers are always
+ * exclusive/inclusive in Kafka/Kinesis indexing service, respectively.
+ *
+ * To be backward compatible with both Kafka and Kinesis datasource metadata when
+ * serializing and deserializing json, redundant constructor fields stream, topic,
+ * partitionSequenceNumberMap and partitionOffsetMap are created. Only one of topic, stream
+ * should have a non-null value and only one of partitionOffsetMap and partitionSequenceNumberMap
+ * should have a non-null value.
+ *
+ * Redundant getters are used for proper Jackson serialization/deserialization when processing terminologies
+ * used by Kafka and Kinesis (i.e. topic vs. stream)
+ */
+public class SeekableStreamEndSequenceNumbers<PartitionIdType, SequenceOffsetType> implements
+    SeekableStreamSequenceNumbers<PartitionIdType, SequenceOffsetType>
+{
+  // stream/topic
+  private final String stream;
+  // partitionId -> sequence number
+  private final Map<PartitionIdType, SequenceOffsetType> partitionSequenceNumberMap;
+
+  @JsonCreator
+  public SeekableStreamEndSequenceNumbers(
+      @JsonProperty("stream") final String stream,
+      // kept for backward compatibility
+      @JsonProperty("topic") final String topic,
+      @JsonProperty("partitionSequenceNumberMap")
+      final Map<PartitionIdType, SequenceOffsetType> partitionSequenceNumberMap,
+      // kept for backward compatibility
+      @JsonProperty("partitionOffsetMap") final Map<PartitionIdType, SequenceOffsetType> partitionOffsetMap
+  )
+  {
+    this.stream = stream == null ? topic : stream;
+    this.partitionSequenceNumberMap = partitionOffsetMap == null ? partitionSequenceNumberMap : partitionOffsetMap;
+
+    Preconditions.checkNotNull(this.stream, "stream");
+    Preconditions.checkNotNull(this.partitionSequenceNumberMap, "partitionIdToSequenceNumberMap");
+  }
+
+  public SeekableStreamEndSequenceNumbers(
+      final String stream,
+      final Map<PartitionIdType, SequenceOffsetType> partitionSequenceNumberMap
+  )
+  {
+    this(stream, null, partitionSequenceNumberMap, null);
+  }
+
+  public SeekableStreamStartSequenceNumbers<PartitionIdType, SequenceOffsetType> asStartPartitions(
```

Comment:
Could you add a javadoc explaining the usage of this method?

Reply:
Added, thanks.